raiden_libs/network/matrix/client.py

```diff
@@ -21,31 +22,53 @@ class GMatrixHttpApi(MatrixHttpApi):
     A wrapper around MatrixHttpApi to limit the number
     of concurrent requests we make to the number of connections
     available to us in requests.Session connection pool size.
+
+    Args:
+        pool_maxsize: max size of underlying/session connection pool
+        retry_timeout: for how long should a single request be retried if it errors
+        retry_delay: callable which returns an iterable of delays
     """
     def __init__(
             self,
             *args,
-            max_retries: int = 3,
             pool_maxsize: int = 10,
+            retry_timeout: int = 60,
```

Comment:
btw, this argument is unecessary since the `retry_delay` is an iteration. You could just exhaust the iterator and then that would be the timeout.

Reply:
No, no, I really want these things separate: the iterator only accounts for the delays **between requests**, it doesn't know how much time may have been spent inside a request. `/sync` requests can take like 30s, if I only count time on the delay generators, this can stack up to several minutes timeout for a single request, and I want that to be predictible (the invariant here is that the raised issue either is solved in a fixed/predictible timeframe, or we give up). A stateful generator to consider startup times will be way more complex than the current ones, while not giving any improvement, as `retry_timeout` is intuitive enough.

Reply:
Also, if we require every client to implement the timeout inside the generator, it may be too easy to forget and have infinite loops in requests which won't be solved. The interface demands that the request either succeeds or raises, so it must be enforced in the interface, with these parameters helping change its behavior but not breaking the expected/eventual possible outcomes.

Reply:
> as retry_timeout is intuitive enough

not true:

```
def once():
  yield 1

GMatrixHttpApi(retry_timeout=60, retry_delay=once)
```

> A stateful generator to consider startup times will be way more complex than the current ones

It just moves complexity around, it doesn't increases it, and it's more correct.

Reply:
With current interface, you **may** enforce the timeout in the delay generator if you want, as you did (by `yield`ing only once and then finishing the generator), but I don't want to **require** that to be the only way to timeout these retries, as it's too easy to forget, so it must be enforced in the interface by default, and optionally implemented like you did, **iff** the user knows what he's doing (by implementing a limited generator, and putting a large-enough `retry_timeout`).